server/src/main/java/org/elasticsearch/common/compress/DeflateCompressor.java

```diff
@@ -88,56 +142,64 @@ public StreamInput streamInput(StreamInput in) throws IOException {
-        decompressedIn = new BufferedInputStream(decompressedIn, BUFFER_SIZE);
-        return new InputStreamStreamInput(decompressedIn) {
-            final AtomicBoolean closed = new AtomicBoolean(false);
-
+        final ReleasableReference<Inflater> current = inflaterForStreamRef.get();
+        final Releasable releasable;
+        final Inflater inflater;
+        if (current.inUse) {
+            // Nested de-compression streams should not happen but we still handle them safely by using a fresh Inflater
+            inflater = new Inflater(true);
+            releasable = inflater::end;
+        } else {
+            inflater = current.get();
+            releasable = current;
+        }
+        InputStream decompressedIn = new InflaterInputStream(in, inflater, BUFFER_SIZE) {
+            @Override
             public void close() throws IOException {
                 try {
                     super.close();
                 } finally {
-                    if (closed.compareAndSet(false, true)) {
-                        // important to release native memory
-                        inflater.end();
-                    }
+                    releasable.close();
                 }
             }
         };
+        return new InputStreamStreamInput(new BufferedInputStream(decompressedIn, BUFFER_SIZE));
     }
 
     @Override
     public StreamOutput streamOutput(OutputStream out) throws IOException {
         out.write(HEADER);
-        final boolean nowrap = true;
-        final Deflater deflater = new Deflater(LEVEL, nowrap);
+        final ReleasableReference<Deflater> current = deflaterForStreamRef.get();
+        final Releasable releasable;
+        final Deflater deflater;
+        if (current.inUse) {
+            // Nested compression streams should not happen but we still handle them safely by using a fresh Deflater
+            deflater = new Deflater(LEVEL, true);
+            releasable = deflater::end;
+        } else {
+            deflater = current.get();
+            releasable = current;
+        }
         final boolean syncFlush = true;
-        DeflaterOutputStream deflaterOutputStream = new DeflaterOutputStream(out, deflater, BUFFER_SIZE, syncFlush);
-        OutputStream compressedOut = new BufferedOutputStream(deflaterOutputStream, BUFFER_SIZE);
-        return new OutputStreamStreamOutput(compressedOut) {
-            final AtomicBoolean closed = new AtomicBoolean(false);
-
+        DeflaterOutputStream deflaterOutputStream = new DeflaterOutputStream(out, deflater, BUFFER_SIZE, syncFlush) {
+            @Override
             public void close() throws IOException {
                 try {
                     super.close();
                 } finally {
-                    if (closed.compareAndSet(false, true)) {
-                        // important to release native memory
-                        deflater.end();
-                    }
+                    releasable.close();
```

Comment:
No need for this kind of check, the `BufferedOutputStream` is a filter output stream that will only close once.

Reply:
Let's add this as a code comment